c_include/fine.hpp

```diff
@@ -1056,6 +1066,34 @@ inline int load(ErlNifEnv *env, void **, ERL_NIF_TERM) {
 }
 } // namespace __private__
 
+// Hash
+namespace __private__ {
+template <> struct Hasher<Atom> {
+  static std::uint64_t hash(HashAlgorithm algorithm, const Atom &atom,
+                            std::uint64_t salt = 0) noexcept {
+    return enif_hash(static_cast<ErlNifHash>(algorithm), *atom.term, salt);
+  }
+};
+
+template <> struct Hasher<Term> {
+  static std::uint64_t hash(HashAlgorithm algorithm, const Term &term,
+                            std::uint64_t salt = 0) noexcept {
+    return enif_hash(static_cast<ErlNifHash>(algorithm), term, salt);
+  }
+};
+} // namespace __private__
+
+template <HashAlgorithm A = HashAlgorithm::PHASH2, typename T>
+inline static std::uint64_t hash(const T &value, std::uint64_t salt = 0) {
+  return __private__::Hasher<T>::hash(A, value, salt);
+}
```

Comment:
I don't think we should be re-exposing `enif_` C APIs, unless it's something very specific to fine, e.g. `fine::make_resource_binary` accepting `fine::ResourcePtr`.

Whenever possible, we should keep the API surface small, and let users use existing `enif_` functions, which they may be already familiar with.

I am in favour of specializing `std::hash`, to enable `std::unordered_map`.

Reply:
Agreed.